### ISSUE TITLE
Development is not public

### ISSUE THREAD

**Issue by thardeck** [keepassium/KeePassium#26]
I am looking forward to KeePassium's release but in my opinion the development is not public.

The commits are not atomic and a huge combination of different changes so it is hard to follow and track. This setup also does make it harder for others to contribute to the project.

Not to mention that KeePassium is a security sensitive open-source software but you seem to have a private repo or separate development infrastructure and only commit the final stages in this repo.

Why are you not using this repository for the every day development if you really want KeePassium to be open-source?

**Comment by keepassium** [comment]
You are absolutely right: the development is not public. 

KeePassium is a personal project. I am personally responsible for 100% of its development and prefer to keep it that way. The motivation behind this is stated in the first lines of the readme:

> KeePassium is a commercial open source app. The "commercial" part gives you a well-maintained app with premium support. The "open source" part gives you the transparency expected from a password manager: you can personally audit the code and build the app you can trust.

Initially, KeePassium was [planned](https://www.reddit.com/r/KeePass/comments/9ow1q1/any_keepass_clients_for_ios_compatible_with_kdbx_4/e7zpzgq/) as a proprietary app. I did not want to invest months of my life only to see somebody grabbing the code and publishing the app for free (this almost happened with [KeePassB](https://github.com/anmipo/keepassb)). However, releasing the source code was the only possible way to [demonstrate](https://www.reddit.com/r/privacytoolsIO/comments/8trrke/provable_privacy_of_a_password_manager/) clean intentions and make the app credible.

Eventually, I had to choose between transparency and commercial safety. That [was not easy](https://www.reddit.com/r/KeePass/comments/a7sa0k/awesome_ios_keepass_app/ec5z0h3) and evolved through most bizarre stages. In the end, I took the risk and [chose transparency](https://www.reddit.com/r/KeePass/comments/bn0rg2/upcoming_ios_port_keepassium/en1hlsf/).

So, the purpose of this repository is to provide transparency. It is not intended to facilitate contributions or derivative work. That's why the commits are squashed, and comments and unit tests are scrubbed — all these factors do not affect the transparency, but improve project's chances of staying afloat in five years.

_P.S. As a side note, I would be happy to make commits more fine-grained. But that would require applying a bunch of fork-deterring steps (mentioned above) to every single commit in `dev` before merging it to `master`. If anyone can explain how to do that, I will be happy to oblige._

**Comment by thardeck** [comment]
> Initially, KeePassium was [planned](https://www.reddit.com/r/KeePass/comments/9ow1q1/any_keepass_clients_for_ios_compatible_with_kdbx_4/e7zpzgq/) as a proprietary app. I did not want to invest months of my life only to see somebody grabbing the code and publishing the app for free (this almost happened with [KeePassB](https://github.com/anmipo/keepassb)).

I can understand the intention but does the Apple Appstore not forbid these kind of forks?
So even if people would take your code, if you have already published your app in the AppStore, in theory Apple would decline the clones so they can not monetize it.
Unless of course they change a lot of things but I am not sure how strict Apple is there.

>However, releasing the source code was the only possible way to demonstrate clean intentions and make the app credible.

This is a general issue with the Apple Appstore because the builds are not reproducible or verifiable. I think it would have been enough if iOS would allow to restrict Internet access for apps, like it is possible for Android, so there is no way your or any other app can leak data without the user noticing.

In this case your app would have been much more credible then parts of it being open-source.

> So, the purpose of this repository is to provide transparency. It is not intended to facilitate contributions or derivative work. That's why the commits are squashed, and comments and unit tests are scrubbed — all these factors do not affect the transparency, but improve project's chances of staying afloat in five years.

At the same time this is a one man project so if your priorities change KeePassium is gone.

Of course this can happen to a publicly developed open-source tool like MiniKeePass too since the developers do not seem to have passed the access over to other interested parties but at least there is the option to take the code and release it under a new name.

And do not get me wrong, it is not about money. I have donated 12 Euros to MiniKeePass in the past - because it was a great tool for my use case and I would have also bought it - but sadly the app seems to be gone for good.

You could add a policy though that if you do not want to continue working on this project anymore, that you publish the whole source code or something like this. This would make it easier to buy the app I think, especially if it is more expensive.

Anyway I wish you good luck with the project and will check it out as soon as it is released in the appstore.

**Comment by keepassium** [comment]
> I can understand the intention but does the Apple Appstore not forbid these kind of forks?
> So even if people would take your code, if you have already published your app in the AppStore, in theory Apple would decline the clones so they can not monetize it.

Yes, they forbid GPL forks on AppStore. Discovering this in October 2018 was exactly [the nudge I needed](https://www.reddit.com/r/KeePass/comments/9ow1q1/any_keepass_clients_for_ios_compatible_with_kdbx_4/e7zj44b/) to open-source the app.

But this is just theory. In practice, I will have to monitor the AppStore for copycats and try to recognize them behind the modified UI. Even worse, I would have to prove to Apple that a vaguely similar proprietary app contains my GPL-licensed code. Lawyer fees, stress... A closed-source project would have been much, *much* easier to protect and commercialize.


> I think it would have been enough if iOS would allow to restrict Internet access for apps, like it is possible for Android, so there is no way your or any other app can leak data without the user noticing.

Yes, this would have made things so much easier: the app cannot leak anything, so you don't have to trust the developer. Unfortunately, this is not a magic bullet:

- If an app can open URLs, it can leak data via GET requests. For example, the app can open its changelog in the system browser. This is a bad user experience (need to switch back), but nothing too suspicious for most users.
- If an app can write files to shared storage/container, these files can be leaked by a companion app that has internet access.


> In this case your app would have been much more credible then parts of it being open-source.

I would disagree with the "parts" part of your sentence. There are no missing assets or precompiled binary blobs. Even third-party modules are directly included in the code. As it is, this repository compiles into a fully-functional KeePassium app.

As for credibility, many people are perfectly happy with the "open source" badge. A few experts will actually audit and compile the code. The remaining few will not be happy with the badge nor willing to read the code. I don't really want to convince them — I would rather focus on building an *incredible* app (pun intended).


> At the same time this is a one man project so if your priorities change KeePassium is gone.

Yeah, this is a frequent argument... against Bitwarden. But there is one important difference: KeePassium does not need servers. (MiniKeePass still runs just fine, right?)

I can see why MiniKeePass' priorities changed: the app was free-as-in-beer. For KeePassium, being a paid app is the best guarantee that it will be maintained. If the app is profitable and my priorities change, it seems fairly simple to hire someone else to work on the project. Otherwise, one just cancels their subscription and stops supporting a dead app.

If I get run over by a bus, there will be nobody to complain about eventual forks. (Only the copyright holder can file a DMCA takedown request.)

> Of course this can happen to a publicly developed open-source tool like MiniKeePass too since the developers do not seem to have passed the access over to other interested parties but at least there is the option to take the code and release it under a new name.

Not really. Nobody may publish MiniKeePass fork on AppStore without their explict permission (see the first paragraph, about GPL forks).

> You could add a policy though that if you do not want to continue working on this project anymore, that you publish the whole source code or something like this. 

Well, the last time my priorities changed, I did exactly that: [released a proprietary app under the MIT license](https://forums.crackberry.com/showthread.php?t=950608&p=12167749&viewfull=1#post12167749). However, there is not much point in making this a commitment: people who need convincing are not my target auditory. 

There are plenty of people who are stuck with an ugly or outdated app. They know exactly what they need and how KeePassium delivers that. I make KeePassium for them.

> Anyway I wish you good luck with the project and will check it out as soon as it is released in the appstore.

Thanks!

**Comment by thardeck** [comment]
Thanks for the explanation.

**Comment by keepassium** [comment]
By coincidence, this morning I [needed](https://sourceforge.net/p/keepass/discussion/329220/thread/0f4c09c6b4/) to reference a few lines in KeePass source code, but... could not find its repository. 

A quick search discovered that KeePass — a one-man project, by the way — [does not have a repository](https://sourceforge.net/p/keepass/discussion/329220/thread/b0bb5457/): Dominik simply publishes snapshots of stable versions. So, there's that.

**Comment by rdxmb** [comment]
https://github.com/keepassxreboot/keepassxc :)

**Comment by keepassium** [comment]
BTW, starting from Feb 2021 commits are no longer squashed.

**Comment by rdxmb** [comment]
I see. I do only use KeePassXC - I did not know KeePass is still a project which is maintained.

**Comment by keepassium** [comment]
> he also seems to think that doing so would impose some great burden of maintenance. 

Well, I kinda understand the "burden" part. Keeping a pristine public repo does take some time, especially for someone with mainly solo experience (speaking from my solo experience here :) 

Merging a PR can easily take a couple of days. After all, a non-trivial PR is more of a suggestion, a shortest-path solution. The maintainer is still responsible for review, refactoring, considering edge cases, code style and testing. Plus, there is time pressure:
- either merge the incoming PR quickly (dropping whatever I was working on),
- or postpone, finish my stuff, and spend some time resolving merge conflicts later. 

So the burden of maintenance is real...
